docs/config.rst
.. _config:

:mod:`~baldaquin.config` --- Configuration
==========================================

The module provides facilities to create, modify, read and write configuration
objects. The basic ideas behind the mechanism implemented here is that:

* configurations are split into sections; more specifically, top-level configuration
  objects are instances of the :class:`Configuration <baldaquin.config.Configuration>`
  class, and each section is an instance of a class inheriting from the abstract
  base class :class:`ConfigurationSectionBase <baldaquin.config.ConfigurationSectionBase>`.
* the configuration section contains a full set of default values for the configuration
  parameters, so that any instance of a configuration object is guaranteed to
  be valid at creation time---and, in addition, to remain valid as the parameter
  values are updated through the lifetime of the object;
* type consistency is automatically enforced whenever a parameter is set or
  updated;
* a minimal set of optional constraints can be enforced on any of the parameters;
* a configuration object can be serialized/deserialized in JSON format so that
  it can be written to file, and the parameter values can be updated from file.

In the remaining of this section we shall see how to declare, instantiate and
interact with concrete configuration objects.


Declaring configuration sections
--------------------------------

The module comes with a number of pre-defined configuration sections that are
generally useful. Basically, all you have to do is to inherit from the abstract
base class :class:`ConfigurationSectionBase <baldaquin.config.ConfigurationSectionBase>`
and override the ``TITLE`` and ``_PARAMETER_SPECS`` top-level class members.

.. literalinclude:: ../src/baldaquin/config.py
   :pyobject: MulticastConfigurationSection

.. literalinclude:: ../src/baldaquin/config.py
   :pyobject: BufferingConfigurationSection

.. literalinclude:: ../src/baldaquin/config.py
   :pyobject: LoggingConfigurationSection

The ``TITLE`` thing is pretty much self-explaining; ``PARAMETER_SPECS`` is
supposed to be an iterable of tuples matching the
:class:`ConfigurationParameter <baldaquin.config.ConfigurationParameter>`
constructor, namely:

1. the parameter name (``str``);
2. the parameter type (``type``);
3. the default value---note its type should match that indicated by the previous field;
4. a string expressing the intent of the parameter;
5. an optional string indicating the physical units of the parameter value (optional);
6. an optional format string for the preferred text rendering of the value (optional);
7. an optional dictionary encapsulating the constraints on the possible parameter
   values (optional).

The rules for parsing the specifications are simple: if the last element of the
tuple is a dictionary we take it as containing all the keyword arguments for the
corresponding parameters; all the other elements are taken to be positional arguments,
in the order specified above. (Note only the first four elements are mandatory.)

The supported constraints are listed in the ``_VALID_CONSTRAINTS`` class variable
of the :class:`ConfigurationParameter <baldaquin.config.ConfigurationParameter>`
class, and they read:

.. literalinclude:: ../src/baldaquin/config.py
   :start-after: # Start definition of valid constraints.
   :end-before: # End definition of valid constraints.

(And if you look this closely enough you will recognize that the constraints
are designed so that the map naturally to the GUI widgets that might be used
to control the configuration, e.g., spin boxes for integers, and combo boxes for
strings to pulled out of a pre-defined list.) Constraints are generally checked and
runtime, and a ``RuntimeError`` is raised if any problem occurs.



Declaring configuration objects
-------------------------------

You can simply take it from here, and declare fully-fledged configuration objects
by just instantiating the :class:`Configuration <baldaquin.config.Configuration>`
class, passing the configuration sections you want to include as arguments.

That said, given how baldaquin is structured, your configuration will likely
contain a number of sections that are common to all the applications (e.g., logging,
buffering, multicasting) and one or more section that is specific to the particular
user application at hand. This common use case is handled by the
:class:`UserApplicationConfiguration <baldaquin.config.UserApplicationConfiguration>`
class, which is again a base class for usable application configuration objects.

By default it comes with no parameters in the user application section, but you
can change that by overriding the ``_PARAMETER_SPECS`` class variable, just like
you did for the configuration sections.

Once you have a concrete class defined, you can instantiate an object, which will
come up set up and ready to use, with all the default parameter values.

>>> config = UserApplicationConfiguration()
>>> print(config)
----------Logging-------------
terminal_level...... INFO
file_enabled........ True
file_level.......... INFO
----------Buffering-----------
max_size............ 1000000
flush_size.......... 100
flush_timeout....... 10.000 s
----------Multicast-----------
enabled............. False
ip_address.......... 127.0.0.1
port................ 20004
----------User Application----

Programmatically, you can retrieve the value of a specific parameter through the
:meth:`value() <baldaquin.config.ConfigurationSectionBase.value()>` class method, and update
the value with :meth:`set_value() <baldaquin.config.ConfigurationSectionBase.set_value()>`.


The :meth:`save() <baldaquin.config.Configuration.save()>` method allows to
dump the (JSON-encoded) content of the configuration into file looking like

.. code-block:: json

  {
      "Logging": {
          "terminal_level": "INFO",
          "file_enabled": true,
          "file_level": "INFO"
      },
      "Buffering": {
          "max_size": 1000000,
          "flush_size": 100,
          "flush_timeout": 10.0
      },
      "Multicast": {
          "enabled": false,
          "ip_address": "127.0.0.1",
          "port": 20004
      },
      "User Application": {}
  }

and this is the basic mechanism through which applications will interact with configuration objects,
with :meth:`update_from_file() <baldaquin.config.Configuration.update_from_file()>`
allowing to update an existing configuration from a JSON file with the proper format.

.. note::

   Keep in mind that configurations are never `read` from file---they come
   to life with all the parameters set to their default values, and then they
   can be `updated` from a JSON file.

   When you think about, this makes extending and/or modifying existing
   configurations much easier as, once the concrete class is changed, all
   existing configuration files are automatically updated transparently, and
   in case one edits a file by hand, any mistake will be promptly signaled
   (and corrected) without compromising the validity of the configuration object.


Module documentation
--------------------

.. automodule:: baldaquin.config


docs/release_notes.rst
.. _release_notes:

Release notes
=============


Version 0.10.0 (2025-10-28)
~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Dependency on aptapy bumped to version 0.11.0 (this fixes issues #85 and #87).
* Reset button hidden for simple applications.
* Added an hyperlink to copy the data file path to clipboard in the main window.
* Added a command-line switch to baldaquin start-app to select the underlying
  matplotlib stylesheet.
* Updated documentation.
* Pull requests merged and issues closed:

    - https://github.com/lucabaldini/baldaquin/pull/91
    - https://github.com/lucabaldini/baldaquin/issues/89
    - https://github.com/lucabaldini/baldaquin/issues/88
    - https://github.com/lucabaldini/baldaquin/issues/87
    - https://github.com/lucabaldini/baldaquin/issues/85


Version 0.9.2 (2025-10-25)
~~~~~~~~~~~~~~~~~~~~~~~~~~

* Maximum strip chart length configurable from the main window in the silly_strip
  application.
* Fix for issue #83 (strip chart not updating correctly at the stop run).
* Pull requests merged and issues closed:

    - https://github.com/lucabaldini/baldaquin/pull/84
    - https://github.com/lucabaldini/baldaquin/issues/83


Version 0.9.1 (2025-10-24)
~~~~~~~~~~~~~~~~~~~~~~~~~~

* Extracted common PlasduinoMainWindow and analog application setup into base
  classes in common.py
* Refactored plasduino_tempmonitor and plasduino_pendulumview to inherit shared
  functionality
* Added cursor activation/deactivation methods to the base analog application class
* Pull requests merged and issues closed:

    - https://github.com/lucabaldini/baldaquin/pull/81


Version 0.9.0 (2025-10-24)
~~~~~~~~~~~~~~~~~~~~~~~~~~

* Integration of VerticalCursor from aptapy plotting utilities for interactive data
  exploration. (Now attached to the ``baldaquin_tempmonitor`` and ``silly_strip`` apps).
* Default refresh interval for PlotCanvasWidget objects changed to 500 ms.
* noxfile refactored and cleaned up.
* Added dependencies on nox and sphinxcontrib-programoutput.
* aptapy dependence bumped to version 0.9.2.
* Documentation updated.
* Pull requests merged and issues closed:

    - https://github.com/lucabaldini/baldaquin/pull/80
    - https://github.com/lucabaldini/baldaquin/issues/76
    - https://github.com/lucabaldini/baldaquin/issues/72
    - https://github.com/lucabaldini/baldaquin/issues/71
    - https://github.com/lucabaldini/baldaquin/issues/68


Version 0.8.2 (2025-10-21)
~~~~~~~~~~~~~~~~~~~~~~~~~~

* Small fix for the package logo on PyPI.
* Pull requests merged and issues closed:

    - https://github.com/lucabaldini/baldaquin/pull/79
    - https://github.com/lucabaldini/baldaquin/issues/78


Version 0.8.1 (2025-10-21)
~~~~~~~~~~~~~~~~~~~~~~~~~~

* Package automatically built and published to PyPI on git tag push.
* Pull requests merged and issues closed:

    - https://github.com/lucabaldini/baldaquin/pull/77
    - https://github.com/lucabaldini/baldaquin/issues/75


Version 0.8.0 (2025-10-21)
~~~~~~~~~~~~~~~~~~~~~~~~~~

* Migrate histogram and strip chart functionality from internal baldaquin modules
  to the external aptapy library.
* New silly_strip example application added to demonstrate the strip chart functionality.
* Pull requests merged and issues closed:

    - https://github.com/lucabaldini/baldaquin/pull/74
    - https://github.com/lucabaldini/baldaquin/issues/70


Version 0.7.0 (2025-10-19)
~~~~~~~~~~~~~~~~~~~~~~~~~~

* Adopted src/ layout with baldaquin package moved to src/baldaquin/
* Introduced dynamic versioning with git integration
* Replaced Makefile with nox sessions for development tasks
* Updated all copyright headers to 2025
* Consolidated environment/logging setup into dedicated modules
* Default quote character changes from single to double quote.
* Major clean up and linting.
* Pull requests merged and issues closed:

    - https://github.com/lucabaldini/baldaquin/pull/73
    - https://github.com/lucabaldini/baldaquin/issues/69


*baldaquin 0.6.0 (Wed, 28 May 2025 11:03:01 +0200)*

* Major restructuring of the configuration facilities, which are now multi-section.
* Logging and buffering configuration controlled from the main window.
* Configuration update from file now resilient to format changes.
* Fix for some funky behavior of the ParameterSpinBox class.
* Minor refactoring of the buffer classes, which are now always unbounded.
* Minor plasduino cleanup.
* New TextLine class and associated text serial communication protocol.
* New compilation and upload scheme for the arduino sketeches and associated artifacts.
* Major refactoring of the xnucleo monitoring, and auto-upload functionality implemented.
* Port class renamed as PortInfo.
* New ArduinoSerialInterface and ArduinoEventHandler classes.
* Cleanup, docs and unit tests added.
* Signature of SerialInterface.connect() changed to accept a PortInfo object.
* Merging pull requests
      * https://github.com/lucabaldini/baldaquin/pull/61
      * https://github.com/lucabaldini/baldaquin/pull/62
      * https://github.com/lucabaldini/baldaquin/pull/66
* Issue(s) closed
      * https://github.com/lucabaldini/baldaquin/issues/54
      * https://github.com/lucabaldini/baldaquin/issues/56
      * https://github.com/lucabaldini/baldaquin/issues/57
      * https://github.com/lucabaldini/baldaquin/issues/64
      * https://github.com/lucabaldini/baldaquin/issues/64


*baldaquin 0.5.0 (Mon, 19 May 2025 09:29:50 +0200)*

* Some bugs and rough edges in the plasduino extension have been fixed.
* New generic monitoring application based on arduino + x-nucleo board.
* Merging pull requests
      * https://github.com/lucabaldini/baldaquin/pull/53
      * https://github.com/lucabaldini/baldaquin/pull/59
* Issue(s) closed
      * https://github.com/lucabaldini/baldaquin/issues/50
      * https://github.com/lucabaldini/baldaquin/issues/51


*baldaquin 0.4.0 (Fri, 31 Jan 2025 15:53:59 +0100)*

* Development notes added to the documentation.
* Setup files for Windows added.

src/baldaquin/config.py
# Copyright (C) 2022--2025 the baldaquin team.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""Configuration facilities.
"""

import datetime
import json
from pathlib import Path
from typing import Any

from .env import BALDAQUIN_DATA, BALDAQUIN_ENCODING
from .logging_ import logger


class ConfigurationParameter:

    """Class representing a configuration parameter.

    This is a simple attempt at putting in place a generic configuration mechanism
    where we have some control on the values we are passing along.

    A configuration parameter is fully specified by its name, type and value, and
    When setting the latter, we make sure that the its type matches.
    Additional, we can specify simple conditions on the parameters that are
    then enforced at runtime.

    Arguments
    ---------
    name : str
        The parameter name.

    type_ : type
        The parameter type.

    value : anything
        The parameter value.

    intent : str
        The intent of the parameter, acting as a comment in the corresponding
        configuration file.

    units : str, optional
        The units for the configuration parameter.

    fmt : str, optional
        An optional format string for the preferred rendering the parameter value.

    constraints : dict, optional
        A dictionary containing optional specifications on the parameter value.
    """

    # Do not remove the following comments, they are used by sphinx to generate the
    # documentation, see https://stackoverflow.com/questions/31561895
    # Start definition of valid constraints.
    _VALID_CONSTRAINTS = {
        int: ("choices", "step", "min", "max"),
        float: ("min", "max"),
        str: ("choices",)
    }
    # End definition of valid constraints.

    def __init__(self, name: str, type_: type, value: Any, intent: str,
                 units: str = None, fmt: None = None, **constraints) -> None:
        """Constructor.
        """
        self.name = name
        self.type = type_
        self.value = None
        self.intent = intent
        self.units = units
        self.fmt = fmt
        for key in tuple(constraints):
            if key not in self._VALID_CONSTRAINTS.get(self.type, ()):
                raise RuntimeError(f"Invalid spec ({key}) for {self.name} ({self.type})")
        self.constraints = constraints
        self.set_value(value)

    def _check_range(self, value: Any) -> None:
        """Generic function to check that a given value is within a specified range.
        """
        if "min" in self.constraints and value < self.constraints["min"]:
            raise RuntimeError(f"Value {value} is too small for {self}")
        if "max" in self.constraints and value > self.constraints["max"]:
            raise RuntimeError(f"Value {value} is too large for {self}")

    def _check_choices(self, value: Any) -> None:
        """Generic function to check that a parameter value is within the
        allowed choices.
        """
        if "choices" in self.constraints and value not in self.constraints["choices"]:
            raise RuntimeError(f"Unexpected choice {value} for {self}")

    def _check_step(self, value: int) -> None:
        """Generic function to check the step size for an integer.
        """
        delta = value - self.constraints.get("min", 0)
        if "step" in self.constraints and delta % self.constraints["step"] != 0:
            raise RuntimeError(f"Invalid value {value} for {self}")

    def set_value(self, value: Any) -> None:
        """Set the parameter value.

        Note that this is where all the runtime checking is performed.
        """
        # Make sure that the value we are passing is of the right type.
        if not isinstance(value, self.type):
            raise RuntimeError(f"Invalid type {value} ({type(value).__name__}) for {self}")
        # Make all the necessary checks on the constraints, if necessary.
        if self.constraints:
            if self.type is int:
                self._check_choices(value)
                self._check_range(value)
                self._check_step(value)
            elif self.type is float:
                self._check_range(value)
            elif self.type is str:
                self._check_choices(value)
        # And if we made it all the way to this point we're good to go :-)
        self.value = value

    def formatted_value(self) -> str:
        """Return the formatted parameter value (as a string).
        """
        if self.fmt is None:
            return f"{self.value}"
        return f"{self.value:{self.fmt}}"

    def pretty_print(self) -> str:
        """Return a pretty-printed string representation of the parameter.
        """
        text = f"{self.name:.<30} {self.formatted_value()}"
        if self.units:
            text += f" {self.units}"
        return text

    def __str__(self):
        """String formatting.
        """
        return f'{self.__class__.__name__} "{self.name}" ({self.type.__name__}, {self.constraints})'


class ConfigurationSectionBase(dict):

    """Base class for a single section of a configuration object. (Note this class
    is not meant to be instantiated, but rather subclassed, overriding the proper
    class variables as explained below.)

    The basic idea, here, is that specific configuration classes simply override
    the ``TITLE`` and ``_PARAMETER_SPECS`` class members, the latter encoding
    the name, type and default values for all the configuration parameters, as well
    as optional help strings and constraints.

    The class interface is fairly minimal, with support to set, and retrieve
    parameter values, and for string formatting.
    """

    TITLE = None
    _PARAMETER_SPECS = ()

    def __init__(self) -> None:
        """Constructor.
        """
        super().__init__()
        # Parse the parameters specs. The basic rule, here, is that if the last
        # element in the spec tuple is a dictionary, we are interpreting it the
        # map of constraints for the parameter, which is otherwise empty.
        # Everything else represent the constructor arguments, in order.
        for specs in self._PARAMETER_SPECS:
            if isinstance(specs[-1], dict):
                args, constraints = specs[:-1], specs[-1]
            else:
                args, constraints = specs, {}
            parameter = ConfigurationParameter(*args, **constraints)
            self[parameter.name] = parameter

    def set_value(self, parameter_name, value) -> None:
        """Update the value of a configuration parameter.
        """
        try:
            # Note all the runtime checking happens in ConfigurationParameter.set_value()
            self[parameter_name].set_value(value)
        except KeyError as exc:
            raise RuntimeError(f'Unknown parameter "{parameter_name}" for '
                               f'{self.__class__.__name__} "{self.TITLE}"') from exc

    def value(self, parameter_name) -> Any:
        """Return the value for a given parameter.
        """
        return self[parameter_name].value

    def formatted_value(self, parameter_name) -> str:
        """Return the formatted value for a given parameter.
        """
        return self[parameter_name].formatted_value()

    def as_dict(self):
        """Return a view on the configuration in the form of a {name: value} dictionary
        representing the underlying configuration parameters.

        This is used downstream to serialize the configuration and writing it to
        file.
        """
        return {parameter.name: parameter.value for parameter in self.values()}

    def __str__(self) -> str:
        """String formatting.
        """
        data = "".join(f"{param.pretty_print()}\n" for param in self.values())
        return f"---------------{self.TITLE:-<25}\n{data}"


class LoggingConfigurationSection(ConfigurationSectionBase):

    """Configuration section for the logging.
    """

    TITLE = "Logging"
    _LOGGING_LEVELS = ("TRACE", "DEBUG", "INFO", "SUCCESS", "WARNING", "ERROR", "CRITICAL")
    _PARAMETER_SPECS = (
        ("terminal_level", str, "DEBUG", "Terminal logging level", dict(choices=_LOGGING_LEVELS)),
        ("file_level", str, "DEBUG", "File logging level", dict(choices=_LOGGING_LEVELS))
    )


class BufferingConfigurationSection(ConfigurationSectionBase):

    """Configuration section for the packet buffering.
    """

    TITLE = "Buffering"
    _PARAMETER_SPECS = (
        ("flush_size", int, 100, "Flush size", dict(min=1)),
        ("flush_timeout", float, 10., "Flush timeout", "s", ".3f", dict(min=1.))
    )


class MulticastConfigurationSection(ConfigurationSectionBase):

    """Configuration section for the packet multicasting.
    """

    TITLE = "Multicast"
    _PARAMETER_SPECS = (
        ("enabled", bool, False, "Enable multicast"),
        ("ip_address", str, "127.0.0.1", "IP address"),
        ("port", int, 20004, "Port", dict(min=1024, max=65535))
    )


class Configuration(dict):

    """Class describing a configuration object, that is, a dictionary of
    instances of ConfigurationSectionBase subclasses.

    Configuration objects provide file I/O through the JSON protocol. One
    important notion, here, is that configuration objects are always created
    in place with all the parameters set to their default values, and then updated
    from a configuration file. This ensures that the configuration is always
    valid, and provides an effective mechanism to be robust against updates of
    the configuration structure.
    """

    def __init__(self, *sections: ConfigurationSectionBase) -> None:
        """Constructor.
        """
        super().__init__()
        for section in sections:
            self.add_section(section)

    def add_section(self, section: ConfigurationSectionBase) -> None:
        """Add a section to the configuration.
        """
        self[section.TITLE] = section

    def update_from_file(self, file_path: str) -> None:
        """Update the configuration dictionary from a JSON file.

        Note we try and catch here all the possible exceptions while updating a
        file, and if anything happens during the update we create a timestamped
        copy of the original file so that the thing can be debugged at later time.
        The contract is that the update always proceeds to the end, and all the
        fields that can be legitimately updated get indeed updated.
        """
        logger.info(f"Updating configuration from {file_path}...")
        with open(file_path, encoding=BALDAQUIN_ENCODING) as input_file:
            data = json.load(input_file)
        errors = False
        for title, section_data in data.items():
            try:
                section = self[title]
            except KeyError:
                logger.warning(f'Unknown configuration section "{title}"')
                errors = True
                continue
            for parameter_name, value in section_data.items():
                try:
                    section.set_value(parameter_name, value)
                except RuntimeError as exc: # noqa: PERF203
                    logger.warning(exc)
                    errors = True
        if errors:
            timestamp = datetime.datetime.now().strftime("%Y%m%d-%H%M%S.%f")
            file_name = f"{Path(file_path).name}.backup.{timestamp}"
            dest = BALDAQUIN_DATA / file_name
            logger.warning(f"Error(s) during update, copying original file to {dest}...")

    def as_dict(self) -> dict:
        """Return a view on the configuration in the form of a dictionary that
        can be used for serialization.
        """
        return {title: section.as_dict() for title, section in self.items()}

    def to_json(self, indent: int = 4) -> str:
        """Encode the configuration into JSON to be written to file.
        """
        return json.dumps(self.as_dict(), indent=indent)

    def save(self, file_path: str) -> None:
        """Dump the configuration dictionary to a JSON file.
        """
        logger.info(f"Writing configuration dictionary to {file_path}...")
        with open(file_path, "w", encoding=BALDAQUIN_ENCODING) as output_file:
            output_file.write(self.to_json())

    def __str__(self):
        """String formatting.
        """
        text = "".join(f"{section}" for section in self.values())
        return text.strip("\n")


class UserApplicationConfiguration(Configuration):

    """Base class for a generic user application configuration.
    """

    _USER_APPLICATION_SECTION_TITLE = "User Application"
    _PARAMETER_SPECS = ()

    def __init__(self) -> None:
        """Constructor.
        """
        super().__init__()
        self.add_section(LoggingConfigurationSection())
        self.add_section(BufferingConfigurationSection())
        self.add_section(MulticastConfigurationSection())

        class _UserApplicationConfigurationSection(ConfigurationSectionBase):

            TITLE = self._USER_APPLICATION_SECTION_TITLE
            _PARAMETER_SPECS = self._PARAMETER_SPECS

        self.add_section(_UserApplicationConfigurationSection())

    def overwrite_section(self, section: ConfigurationSectionBase) -> None:
        """Overwrite a section in the configuration.
        """
        if section.TITLE not in self:
            raise RuntimeError(f'Unknown section "{section.TITLE}" for '
                               f'{self.__class__.__name__}')
        self[section.TITLE] = section

    def logging_section(self):
        """Return the logging section of the configuration.
        """
        return self[LoggingConfigurationSection.TITLE]

    def buffering_section(self):
        """Return the buffering section of the configuration.
        """
        return self[BufferingConfigurationSection.TITLE]

    def multicast_section(self):
        """Return the multicast section of the configuration.
        """
        return self[MulticastConfigurationSection.TITLE]

    def application_section(self):
        """Return the user application section of the configuration.
        """
        return self[self._USER_APPLICATION_SECTION_TITLE]
